runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/graph/GreedyStageFuser.java

```diff
@@ -141,6 +149,55 @@ public static ExecutableStage forGrpcPortRead(
         materializedPCollections);
   }
 
+  /**
+   * Patches all input coders of stateful ExecutableStage transforms. Stateful transforms aways have
+   * a KvCoder as input. Flink partitions the data based on the serialized version of the key. This
+   * key must retain the same binary representation to be able to serve the state in state requests
+   * from the SDK Harness from the correct partition. If the binary representation does not match,
+   * this will result in inconsistent checkpoints because keys are associated to a Flink key group
+   * which does not belong to the key group range of the state handling operator.
+   */
+  private static RunnerApi.Components lengthPrefixKeyCoder(
```

Comment:
Now that I looked through the shared code more, we don't have access to the wire coder at this point in time and really we are instantiating it in https://github.com/apache/beam/blob/6b3f7d62d3a66b750a7bd541a94646fdf735bbd7/runners/java-fn-execution/src/main/java/org/apache/beam/runners/fnexecution/control/ProcessBundleDescriptors.java#L174

There we would have access to the exact coder id that we should be setting on the main input pcollection.

That class seems like the appropriate place to do this modification and it is also the place where we instantiate the runner coder related to the user state specs.

Sorry that I didn't realize this earlier when I made my suggestion to place it in the GreedyStageFuser.

Reply:
I agree that `ProcessBundleDescriptors` is the better place. No need to modify the executable stage up front.

Reply:
Yes, this looks much cleaner now.